### ISSUE TITLE
S3 backend throwing 500 on versioned bucket

### ISSUE THREAD

**Issue by thehesiod** [spulec/moto#5588]
This is an injection that happened between 4.0.5 + 4.0.8.  The issue is due to line https://github.com/spulec/moto/blob/master/moto/s3/models.py#L1616 which is doing a deep copy of something it should not be deep copying as the structure is complicated with locks and what not.  In certain scenarios the file will be closed causing it to throw an exception.  See https://github.com/spulec/moto/issues/5341#issuecomment-1287429197

**Comment by bblommers** [comment]
Hi @thehesiod, I can't reproduce this, I'm afraid. This is my attempt:
```
import asyncio
import aiobotocore
import boto3
import botocore
import itertools
from aiobotocore.session import get_session
from moto import mock_s3


class PatchedAWSResponse:
    def __init__(self, response: botocore.awsrequest.AWSResponse):
        self._response = response
        self.status_code = response.status_code
        self.raw = response.raw
        self.raw.raw_headers = {}

    @property
    async def content(self):
        return self._response.content

def factory(original):
    def patched_convert_to_response_dict(http_response, operation_model):
        return original(PatchedAWSResponse(http_response), operation_model)

    return patched_convert_to_response_dict

aiobotocore.endpoint.convert_to_response_dict = factory(aiobotocore.endpoint.convert_to_response_dict)
empty_tuple = ()

@mock_s3
def test():
    loop = asyncio.get_event_loop()
    loop.run_until_complete(deleteall())
    loop.close()


async def deleteall():
    session = get_session()

    async with session.create_client('s3', region_name='us-east-1') as client:
        bucket = "bucketx"
        await client.create_bucket(Bucket=bucket)
        await client.put_bucket_versioning(Bucket=bucket, VersioningConfiguration={'Status': 'Enabled'})

        for i in range(100):
            await client.put_object(Bucket=bucket, Key=f"key{i}", Body="sth")

        paginator = client.get_paginator('list_object_versions')
        async for page in paginator.paginate(Bucket=bucket):
            for s3_obj in itertools.chain(page.get('Versions', empty_tuple), page.get('DeleteMarkers', empty_tuple)):
                await client.delete_object(Bucket=bucket, Key=s3_obj['Key'], VersionId=s3_obj['VersionId'])
```
That's on master, with werkzeug 2.2.2 and the latest versions of aiobotocore etc.

**Comment by thehesiod** [comment]
btw I'm using the server and not the patch.  Lemme c if I can get a reproable use-case

**Comment by thehesiod** [comment]
hah tricky, simple test case doesn't work

**Comment by bblommers** [comment]
I just saw the stacktrace that you posted - looks like the GC PR's that we merged are the rootcause: https://github.com/spulec/moto/pull/5556/files

TLDR: S3 opens filehandles automatically when creating a `FakeKey()`-object, but we never close them (until now). These PR's were raised to close them explicitly everytime a FakeKey is deleted.

Running an async delete probably looks like this:
 - thread 1: call delete_file
 - thread 1: close file handle on the file
 - thread 2: deep copy that exact file
 - thread 1: remove the file from the bucket
 - thread 2: boom

So Moto should ensure that either 1) we close/delete in a single operation, or 2) we delete first to ensure that no-one can reach it - and then close the filehandle.

**Comment by thehesiod** [comment]
i just saw that it's related to multipart

**Comment by thehesiod** [comment]
boom https://gist.github.com/thehesiod/31504dd47505f38b59b3e7a238370201

**Comment by bblommers** [comment]
Awesome - thanks for putting that together @thehesiod!

**Comment by bblommers** [comment]
This should be fixed as of moto >= 4.0.9.dev5 - can you verify this @thehesiod?

**Comment by thehesiod** [comment]
if my gist works it will work on our system as it's the same code

**Comment by thehesiod** [comment]
thank you for the extremely quick fix! I have a big backlog for aiobotocore :(
